build: add kernel + initramfs to handle in-line updates from old buildroot.

vegano1 opened the pull request:
[RET-1241](https://opentrons.atlassian.net/browse/RET-1241)

**Background**
We can't update the old buildroot to the new buildroot directly because the dtb + overlays have changed for the new kernel. If we try and update directly, the update will go through and get installed on the inactive partition as expected. However, once the device reboots into the new kernel it will hang, effectively bricking the device and requiring an sd card reflash.

To deal with this we are adding "KernelRamfs" which is a small kernel + initramfs that takes the place of the new kernel and bootstraps the update by updating the boot partition with the appropriate files. The update path is as followed

1. When we are updating from pre-6.2, after the rootfs has been written to the inactive partition and the device reboots, the existing u-boot script will look for the kernel in **/boot/zImage**. Since **/boot/zImage** is now a symlink to /boot/kernelRamfs alongside the original pre-6.2 dtb at fdt_addr as set by start.elf, we will boot into kernelRamfs.

2. Once we are in the KernelRamfs, the init script will mount the boot partition **/dev/mmcblk0p1** to /boot and the new rootfs partition **/dev/mmcblk0p(2-3)** to /rootfs. It then verifies that all the required update files (dtb + overlays + config.txt + boot.scr + kernelRamfs.dtb) contained in the /boot/boot.tar.xz file in the mounted rootfs. If the files are present it proceeds to copy them to the boot partition, unmounts the partitions, and reboots the system.

3. Once we restart, the new u-boot script will load the kernel in /boot/kernel from the active rootfs partition determined by the 'ot2part' u-boot env variable. Once we load the kernel we use the device tree blob at fdt_addr and boot into post-6.2 kernel, from here the system boots as expected.

**KernelRamfs Functionality**
There is a new u-boot env flag 'kernelramfs' which determines if we boot into kernelRamfs or the regular kernel, by default this is set to 0 so the system boots into the normal kernel. However, it can be set to a different value from userspace to change the logic as followed.

kernelramfs = "0"   - load kernel at /boot/kernel
kernelramfs = "1"   - load kernelRamfs at /boot/kernelRamfs and drop into recovery shell
kernelramfs > "1"   - load kernelRamfs at /boot/kernelRamfs and update the boot partition with the contents of the /boot/boot.tar.gz file in the active rootfs partition (2,3) as determined by the 'ot2part' u-boot env variable.

**Changes**
- added KernelRamfs to build
- updated boot.scr to handle new boot path
- Trimmed down our rootfs by removing packages "rpi-userland, dtc, screen, tmux, which, non-essential ffmpeg configs (saved ~8mb)
- Trimmed down our actual kernel by removing a bunch of unused modules (saved ~10mb)

**Risk Assessment**
- Very high, we are updating the boot partition which can brick the device.
- Does not support downgrading from 6.2 to lower

**Testing**
- [ ] validate on non-refresh OT2
- [x] update from pre-6.2 passes
- [x] downgrade from 6.2 does not brick the device but instead goes back to 6.2

vegano1 committed:
create minimal kernel

vegano1 committed:
configs/ot2_defconfig: remove unused packages
remove rpi-userland
remove dtc and dtc-programs
remove screen
remove tmux
remove which
trim down ffmpeg for our specific case

vegano1 committed:
linux-kernel: reduced kernel size
removed unused drivers + modules from kernel
optimized kernel for size
remove bootup image
remove debug kernel logging

vegano1 committed:
move tiny-config and related files to its own dir
use initramfs option from buildroot instead of kernel

vegano1 committed:
add seperate output directories /output/tinyKernel and /output/ot2 and pass them into buildroot
add ccache
clean up unused tinyKernel files
use xz compression method for cpio to reduce zImage space

vegano1 committed:
Merge branch 'update-buildroot-2022.02.4' into inline_update_fix

vegano1 committed:
rename to kernalRamfs

vegano1 committed:
add recovery function to jump to shell if something fails
fix fw_printenv by adding fw_env.config and addresing /var/lock & /run/lock missing dirs

vegano1 committed:
added more context for KernelRamfs
update copy source location

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,78 @@
+
+echo "Setting up ramdisk"
+# Mount the /proc, /sys and /dev filesystems.
+/bin/mount -t proc none /proc || recovery
```
might be better to `set -eo pipefail` and `trap recovery EXIT`, saves you from having to do `|| recovery` all over the place (and saves you from forgetting to do that on some random line)

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,78 @@
+# NOTE (ba, 2022-10-16): we need to set the ot2part env variable so 
+# uboot does not revert to the old partition once we restart. We
+# want to boot on the same partition but with updated kernel.
+boot_part=$(grep -o [2-3]) || recovery
```
what's this searching? is it missing a `fw_getenv`?

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,78 @@
+# want to boot on the same partition but with updated kernel.
+boot_part=$(grep -o [2-3]) || recovery
+echo "Setting ot2part to $boot_part"
+fw_setenv ot2part $boot_part || recovery
```
probably make this `/bin/fw_setenv` just to stay consistent

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,78 @@
+echo "Copying files to boot partition..."
+# copy the dtb(s) + overlays + config.txt from rootfs to boot
+cp /rootfs/boot/*.dtb /boot/ || recovery
+cp /rootfs/boot/overlays/overlays/*.dtbo /boot/overlays/ || recovery
```
we should `mkdir -p` these just in case

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,78 @@
+fw_setenv ot2part $boot_part || recovery
+
+echo "Renaming kernel back to zImage"
+# overwrite the kernelRamfs with the new kernel
```
can we instead 
- mount a tmpfs
- `mv /rootfs/boot/zImage /tmp/kernelRamfs ; mv /rootfs/boot/kernel /rootfs/boot/zImage ; mv /tmp/kernelRamfs /rootfs/boot/kernelRamfs`

just so we keep this around, just in case. we're guaranteed to have the space after all.

sfoster1 reviewed (approved):
This is really cool and fundamentally looks good. I think there's some little things we could improve about the initscript but that's about it.

We also might be able to trim the mini image down more by trimming some more stuff out of its busybox I guess, but we'll definitely be shaving kb at that point.

vegano1 committed:
add kernelRamfs.dtb so it can always load
update u-boot boot.scr so we can jump into the kernel, or kernelRamfs if set from userspace or we cant boot
clean up init and let us shell into it for recovery mode;

sfoster1 commented on board/opentrons/ot2/boot.scr:
```diff
@@ -46,17 +55,41 @@ for arg in $fdtargs; do
+        setenv kernelramfs 1
+        sleep 5
+    fi
+    fail_count="$fail_count$fail_count"
```
won't this exponentially double - first it's f; then `$fail_count$fail_count` = `ff`; then `$fail_count` is ff, so `$fail_count$fail_count` is ffff

sfoster1 commented on board/opentrons/ot2/boot.scr:
```diff
@@ -46,17 +55,41 @@ for arg in $fdtargs; do
+    else
+        echo "Loading Opentrons KernelRamfs"
+        # load the device tree from boot partition
+        fatload mmc 0:1 $fdt_addr /kernelRamfs.dtb
```
where does this dtb come from the first boot after an upgrade from pre-6.2?

sfoster1 replied:
i think this has a couple problems, actually:
- If this is the first boot into 6.2 from 6.1 or previous, `kernelRamfs.dtb` won't be in the boot partition, so we won't be able to load it
- If this is a boot into 6.1 after downgrading from 6.2, and we're about to load a 6.1 kernel, then `kernelRamfs.dtb` won't be correct (unless it's the pre-6.2 dtb i guess)

sfoster1 commented on board/opentrons/ot2/kernelRamfs/post-image.sh:
```diff
@@ -0,0 +1,10 @@
+# rename the kernel so it does not get overwritten when the ot2_defconfig gets built
+echo "renaming zImage to kernelRamfs"
+mv ${BINARIES_DIR}/zImage ${BINARIES_DIR}/kernelRamfs
+cp ${BOARD_DIR}/kernelRamfs.dtb ${BINARIES_DIR}
```
why use a checked-in dtb like this rather than compile it?

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,107 @@
+
+recovery() {
+   # drop into recovery shell if we get a non-zero exit status
+   [ $? != 1 ] && exit
```
probably `[ $? == 0 ] && exit` i'd think, `0` is success and _everything else_ is failure typically

sfoster1 commented on board/opentrons/ot2/post-build.sh:
```diff
@@ -24,7 +24,23 @@ EOF
+# copy the KernelRamfs + actual Kernel images
+cp ${BINARIES_DIR}/zImage ${TARGET_DIR}/boot/kernel
+cp ./output/kernelRamfs/images/kernelRamfs ${TARGET_DIR}/boot/zImage
+cp ./output/kernelRamfs/kernelRamfs.dtb /${TARGET_DIR}/boot/
```
```suggestion
cp ./output/kernelRamfs/kernelRamfs.dtb ${TARGET_DIR}/boot/
```

sfoster1 commented on board/opentrons/ot2/boot.scr:
```diff
@@ -4,6 +4,15 @@
 
+# Check if we have an env var for whether we should jump into KernelRamfs
+# or regular kernel, this could be set from userspace but also missing.
+if test -z "$kernelramfs"
```
i think this wants to be something like `kernelramfs_complete` because
- if you are booting for the first time into 6.2, the variable will not be present and should default to
- during an update, update code should set it to 0
- i think that if this variable is 0/not present, we _should_ enter the kernelramfs, since that is what applies boot partition fixups so the regular kernel can boot

sfoster1 replied:
so what it's indicating is actually "I have already performed all necessary fixups to boot a 6.2 kernel"

sfoster1 commented on board/opentrons/ot2/boot.scr:
```diff
@@ -46,17 +55,41 @@ for arg in $fdtargs; do
-    bootz $kernel_addr_r - $fdt_addr
 
+    # Boot into the regular kernel, if 'kernelramfs' is set boot into KernelRamfs
+    if test "$kernelramfs" = "0"
```
under the above assumption about this being `kernelramfs_complete`, if it's `1` then we should boot to `/boot/kernel`, not if it's `0`/not present

sfoster1 commented on board/opentrons/ot2/boot.scr:
```diff
@@ -46,17 +55,41 @@ for arg in $fdtargs; do
+        echo "Loading Opentrons KernelRamfs"
+        # load the device tree from boot partition
+        fatload mmc 0:1 $fdt_addr /kernelRamfs.dtb
+        ext4load mmc 0:$ot2part $kernel_addr_r /boot/kernelRamfs
```
this wants to be `/boot/zImage`, right?

sfoster1 commented on board/opentrons/ot2/boot.scr:
```diff
@@ -46,17 +55,41 @@ for arg in $fdtargs; do
+    then
+        echo "Loading Kernel"
+        ext4load mmc 0:$ot2part $kernel_addr_r /boot/kernel
+    else
```
this part should happen if `kernelramfs_complete` == `0`/not present

sfoster1 replied:
and this path is also the "downgraded to 6.1 from 6.2" path, since it's what loads `zImage`

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,107 @@
+# mount the boot partition
+mount /dev/mmcblk0p1 /boot
+
+# if the 'kernelramfs' env flag does not exist,
```
this would be our `kernelramfs_complete` flag, and we should set it when we're done executing, and we should set it to 1

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,107 @@
+   fw_setenv kernelramfs 0
+fi
+
+# if kernelramfs env flag is set to "1" jump the shell
```
this would happen if
- there was a logic bug in the boot script
- the boot script had been replaced by the old boot script
right? 

If it's case 1, we'd definitely like to know; if it's case 2, then we need to do the same boot partition update we do anyway. I think we should unconditionally do the boot partition update; maybe touch a dummy file in the boot partition to indicate an error; and continue

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,107 @@
+
+echo "Ramdisk setup, starting update"
+# find the active rootfs and mount it
+ROOTFS=$(cat /proc/cmdline | grep -o "/dev/mmcblk0p.")
```
I don't really like this because it might be fragile in the future; I'd prefer something like `grep -oz "rootfs=.*" /proc/cmdline | grep -o "/dev/mmcblk0p."`

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,107 @@
+
+# use ls command to make sure the update files exist
+# redirect output to null so we arent flooded with messages
+# we dont have brace expantion so have to call ls multiple times.
```
```suggestion
# we dont have brace expansion so have to call ls multiple times.
```

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,107 @@
+# use ls command to make sure the update files exist
+# redirect output to null so we arent flooded with messages
+# we dont have brace expantion so have to call ls multiple times.
+ls /boot/*.dtb > /dev/null
```
i don't really get why we'd both be doing this and be doing the `mkdir -p` above. these commands will cause the script to fail and enter recovery; they'll fail both if the directories don't exist and if the directories exist but are empty (because we just created them). it seems redundant.

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,107 @@
+# We need to set the ot2part env variable so uboot does not revert to
+# the old partition once we restart. Since we want to boot on the same
+# partition but with updated kernel.
+boot_part=$(fw_printenv -n ot2part)
```
this really doesn't seem like it would do anything? we're just writing back to the fw env with the same value it used to have.

sfoster1 replied:
it would be necessary if we were terminating the kernel by letting this script exit, but if we're calling `reboot` then the bootloader won't see this as failing.

sfoster1 commented on board/opentrons/ot2/kernelRamfs/rootfs-overlay/init:
```diff
@@ -0,0 +1,107 @@
+   # drop into recovery shell if we get a non-zero exit status
+   [ $? != 1 ] && exit
+   echo "Something went wrong, entering recovery..."
+   setsid sh -c 'exec sh </dev/tty1 >/dev/tty1 2>&1'
```
This is useful for debugging but really not shippable. No customer will be able to see this output; they'll just think the robot failed to boot. What needs to happen instead is that we need to write some log to the boot partition about how we failed and then bail by letting this script exit so the kernel stops because pid1 stopped.

vegano1 replied:
agreed, we will reboot if something goes wrong instead.

sfoster1 commented on board/opentrons/ot2/boot.scr:
```diff
@@ -46,17 +55,41 @@ for arg in $fdtargs; do
+        ext4load mmc 0:$ot2part $kernel_addr_r /boot/kernelRamfs
+    fi
+
+    # boot the correspoing kernel + device tree blob
```
```suggestion
    # boot the corresponding kernel + device tree blob
```

sfoster1 commented on board/opentrons/ot2/post-build.sh:
```diff
@@ -24,7 +24,23 @@ EOF
 mkdir -p ${TARGET_DIR}/boot
-cp ${BINARIES_DIR}/zImage ${TARGET_DIR}/boot/zImage
+
+# NOTE: we want to boot into the KernalRamfs which is why
```
```suggestion
# NOTE: we want to boot into the KernelRamfs which is why
```

sfoster1 commented on board/opentrons/ot2/post-build.sh:
```diff
@@ -24,7 +24,23 @@ EOF
+
+# NOTE: we want to boot into the KernalRamfs which is why
+# we are renaming it to zImage, this way when updating from
+# old kernel to new kernel we will boot into KernalRamfs first.
```
```suggestion
# old kernel to new kernel we will boot into KernelRamfs first.
```

sfoster1 commented on board/opentrons/ot2/post-build.sh:
```diff
@@ -24,7 +24,23 @@ EOF
+# we are renaming it to zImage, this way when updating from
+# old kernel to new kernel we will boot into KernalRamfs first.
+# This will copy the dtb + overlays + config.txt in the rootfs
+# to the boot partition, lastly we rename the actual "kernel" back
```
this is out of date i think

sfoster1 reviewed (changes requested):
Ok, first set of in-depth look at this. Obviously some of this is just because you haven't quite finished yet, but I think we need to slightly change our approach here - see the longer comments in the boot script and the init script

vegano1 committed:
copy kernelRamfs to kernelramfs_bin dir, then completely clean the output dir before doing a regular build
fix boot.scr fail_count iteration

vegano1 committed:
fix dtb

sfoster1 commented on board/opentrons/ot2/boot.scr:
```diff
@@ -4,6 +4,15 @@
+# or regular kernel, this could be set from userspace but also missing.
+if test -z "$kernelramfs"
+then
+    echo "No saved kernelramfs, defaultint to 0"
```
```suggestion
    echo "No saved kernelramfs, defaulting to 0"
```

sfoster1 reviewed (approved):
Okay, I understand this a lot more now. Looks good to me from a code review perspective!

vegano1 committed:
fix zImage symlink to kernelRamfs
update init script so we set the bootdelay=-2 uboot var for non-refresh ot2
add CONFIG_TIMEOUT=10 to kernelRamfs linux defconfig so we restart if we kernel panic
dont saveenv in uboot script when setting kernelramfs=1

vegano1 committed:
fix typo board/opentrons/ot2/boot.scr

Co-authored-by: Seth Foster <seth@opentrons.com>

vegano1 committed:
fix typo board/opentrons/ot2/kernelRamfs/rootfs-overlay/init

Co-authored-by: Seth Foster <seth@opentrons.com>

vegano1 committed:
remove dead dtb
uncomment sdk
